### ISSUE TITLE
[C++] Add software implementation of PDEP and reenable BMI2 code paths on all AVX2 CPUs

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#37623]
### Describe the enhancement requested

In #37610 we had to guard some code paths on the presence of the BMI2 instruction sets, because those code paths use two BMI2 instrinsics (the PDEP and PEXT instructions) that are not efficiently implemented on all modern x86 CPUs (and that can be absent on a few rare ones).

However, we already have a reasonably good software emulation of PEXT [in another place](https://github.com/apache/arrow/blob/a8a604cfc5bc8d05ca3b1ef1388e6efd8ba6e5cd/cpp/src/parquet/level_conversion_inc.h#L37-L257) and it should be relatively easy to write a software emulation of PDEP using the exact same approach (lookup table to process 5 bits at a time).

This would allow re-enabling the aforementioned codepaths on all CPUs supporting AVX2, even if they don't have an (efficient) BMI2 implementation.

This includes AMD CPUs since at least 2017.

### Component(s)

C++

**Comment by pitrou** [comment]
@felipecrv Perhaps you would be interested.

**Comment by pitrou** [comment]
Also @westonpace FYI

**Comment by felipecrv** [comment]
> @felipecrv Perhaps you would be interested.

Yes. Using that software implementation from the AVX2 code was my plan to fix #37610. I will prepare a PR.

**Comment by mapleFU** [comment]
I'm interested in this patch.

@pitrou Also, should I extract "ExtractBits" to some other place? Currently `_pext` is implemented in https://github.com/apache/arrow/blob/b8fff043c6cb351b1fad87fa0eeaf8dbc550e37c/cpp/src/parquet/level_conversion_inc.h#L267 . Should I also extract this?

**Comment by pitrou** [comment]
@mapleFU Yes, solving this issue probably implies to reorganize the code (and potentially hide the PDEP routines behind batched functions, so that the implementation is not inlined in a header file).

**Comment by mapleFU** [comment]
Did we have some place for this? Like `arrow/bit_util.h`? Or I should create two new file for this?

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
